### DISCUSSION TITLE
Multiline PHPDoc types?

### DISCUSSION THREAD

**Discussion by ondrejmirtes** [vimeo/psalm#9969] [General]
Hi,
how does Psalm go about multiline PHPDoc types?

In Doctrine there's: https://github.com/doctrine/orm/blob/4c3bd208018c26498e5f682aaad45fa00ea307d5/lib/Doctrine/ORM/Query/Parser.php#L2569-L2577

And it seems like Psalm is handling it fine, but I'm struggling to do the same thing in the playground: https://psalm.dev/r/de76ee99bc

PHPStan supports newlines in these places:

* In conditional types in few specific places
* In generics `Foo<Bar, Baz>` in few specific places
* In callable signatures in few specific places
* In array shapes in few specific places
* Unions and intersections can be multiline as long as they're parenthesized

Because of phpdoc-parser issues like this https://github.com/slevomat/coding-standard/issues/1586 I want to expand the multiline support, but I need to know exactly how other people do it :)

Thank you.

**Comment by weirdan** [comment]
>  how does Psalm go about multiline PHPDoc types?

In general? We try to support them, but it's not always possible as the PHPDoc tag syntax is ambiguous, and even where it is possible, the support is spotty at best. So I've been recommending users parenthesize types to avoid the ambiguity.

**Reply by ondrejmirtes** [reply]
Alright, so it's not probably worth it trying to make it work with phpstan/phpdoc-parser. Thanks!
